Enable Model and Tokenizer to directly load paddlepaddle models from huggingface hub

sijunhe opened the pull request:
<!-- Demo: https://github.com/PaddlePaddle/PaddleNLP/pull/26 -->
### PR types
New features
<!-- One of [ New features | Bug fixes | Function optimization | Performance optimization | Breaking changes | Others ] -->

### PR changes
APIs
<!-- One of [ Models | APIs | Docs | Others ] -->

### Description
Enable Model and Tokenizer to directly load paddlepaddle models from huggingface hub

#### Test Tokenizer
```
# load from HF works
tokenizer = AutoTokenizer.from_pretrained("PaddlePaddle/ernie-3.0-nano-zh", from_hf_hub=True)
# load official pre-trained works
tokenizer = AutoTokenizer.from_pretrained("ernie-3.0-nano-zh")
tokenizer.save_pretrained("temp_test_tokenizer")
# load from local
tokenizer = AutoTokenizer.from_pretrained("temp_test_tokenizer")
```

#### Test Model
```
# load from HF works
model = AutoModel.from_pretrained("PaddlePaddle/ernie-3.0-nano-zh", from_hf_hub=True)
# load official pre-trained works
model = AutoModel.from_pretrained("ernie-3.0-nano-zh")
model.save_pretrained("temp_test_tokenizer")
# load from local
model = AutoModel.from_pretrained("temp_test_tokenizer")
```

<!-- Describe what this PR does -->

sijunhe committed:
wip

sijunhe committed:
wip

sijunhe committed:
AutoModel and AutoTokenizer works

sijunhe committed:
eady for review

sijunhe committed:
isort and yapf

sijunhe committed:
merge with develop

sijunhe committed:
fix style

wj-Mcat committed:
Merge branch 'develop' into hf_hub_integration

wj-Mcat commented on paddlenlp/transformers/auto/modeling.py:
```diff
@@ -166,10 +170,52 @@ def __init__(self, *args, **kwargs):
+    def _get_model_class_from_config(cls, pretrained_model_name_or_path,
+                                     config_file_path):
+        with io.open(config_file_path, encoding="utf-8") as f:
+            init_kwargs = json.load(f)
+        # class name corresponds to this configuration
+        init_class = init_kwargs.pop("init_class", None)
+        init_class = init_class[:-5] if init_class.endswith(
+            "Model") else init_class
+        if init_class:
+            for model_flag, name in MAPPING_NAMES.items():
+                if model_flag in init_class:
+                    model_name = model_flag + 'Model'
+                    break
+        else:
+            # From pretrained_model_name_or_path
+            for model_flag, name in MAPPING_NAMES.items():
+                if name in pretrained_model_name_or_path.lower():
+                    model_name = model_flag + 'Model'
+                    break
+        init_class = cls._name_mapping[model_name + '_Import_Class']
+        class_name = cls._name_mapping[init_class]
+        import_class = importlib.import_module(
+            f"paddlenlp.transformers.{class_name}.modeling")
+        try:
+            model_class = getattr(import_class, init_class)
+            return model_class
+        except AttributeError as err:
+            logger.error(err)
+            all_model_classes = import_class.__all__
+            all_tasks = {
+                get_task_name(m)
+                for m in all_model_classes if get_task_name(m) is not None
+            }
+            raise AttributeError(
+                f"module '{import_class.__name__}' only supports the following classes: "
+                + ", ".join(m for m in all_model_classes) + "\n"
+                "Hint: you can use interface " +
+                " or ".join(task + ".from_pretrained" for task in all_tasks) +
+                f" to load '{pretrained_model_name_or_path}'\n")
+
```
这个方法在我的`cli` PR里面有，后续此类方法可抽离成utils公共模块。

wj-Mcat replied:
由于此方法比较通用，我建议可放到`paddlenlp/transformers/utils.py`模块当中，这样其他模块也可以复用，你觉得呢？

wj-Mcat replied:
另外，目前为了和 hf 对齐，我们也是兼容：`architectures`这个字段的，所以在这个模块也是考虑此字段的信息的，特别是针对于未来新模型。

sijunhe replied:
out of the scope for this PR but added a TODO

wj-Mcat commented on paddlenlp/transformers/auto/modeling.py:
```diff
@@ -224,48 +270,27 @@ def _from_pretrained(cls,
+            config_file = hf_hub_download(repo_id=pretrained_model_name_or_path,
+                                          filename=cls.model_config_file,
+                                          cache_dir=MODEL_HOME)
+            if os.path.exists(config_file):
+                model_class = cls._get_model_class_from_config(
+                    pretrained_model_name_or_path, config_file)
+                logger.info("We are using %s to load '%s'." %
+                            (model_class, pretrained_model_name_or_path))
+                return model_class.from_pretrained(
+                    pretrained_model_name_or_path,
+                    from_hf_hub=from_hf_hub,
+                    *model_args,
+                    **kwargs)
```
如果`config_file`不存在的话，是否需要给一个 warning 或者 error 呢？

因为如果说是不存在的话，理论上应该终止模型的初始化逻辑处理。

sijunhe replied:
added a warning

sijunhe replied:
added a logger.warning messasge

wj-Mcat commented on paddlenlp/transformers/auto/tokenizer.py:
```diff
@@ -136,7 +140,48 @@ def __init__(self, *args, **kwargs):
+            return tokenizer_class
+        # If no `init_class`, we use pattern recognition to recognize the tokenizer class.
+        else:
+            print(
+                'We use pattern recognition to recognize the Tokenizer class.')
```
这类提示信息都应该基于`logger`来提供给开发者吧，是否需要将`print`改成`logger.info`呢？

wj-Mcat commented on paddlenlp/transformers/auto/tokenizer.py:
```diff
@@ -136,7 +140,48 @@ def __init__(self, *args, **kwargs):
+            print(
+                'We use pattern recognition to recognize the Tokenizer class.')
+            for key, pattern in cls._name_mapping.items():
+                if pattern in pretrained_model_name_or_path.lower():
```
我在想，这里的`pattern in name_or_path`这个操作会不会太松散，比如在以下情况下就会出现问题：

```python
cls._name_mapping = {
    "BertTokenizer": "bert",
    "AlbertTokenizer": "albert",
    "RobertaTokenizer": "roberta",
}
pretrained_model_name_or_path = "tinybert-4l-312d"
```

sijunhe replied:
since this is the original code, it's out of scope for this PR. I added a TODO and referenced your comment.

wj-Mcat reviewed (changes requested):
你的这个 pr 我觉得非常棒，看起来很令人兴奋，不过我除了留的 comment，还有几点想跟你讨论：

1. 你现在是直接在老版本的from_pretrained（新版本为`from_pretrained_v2`）上面做的，可是后面的主流模型都会迁移到新版本上，所以我感觉这部分的调整是否直接在新的接口上来完成是否比较好？
2. `from_hf_hub`的核心逻辑应该是数据源的问题，那是否可以把这部分的逻辑抽离成一个单独的函数来完成下载的这个工作呢？
3. 是否不用添加`from_hf_hub`这个参数，当在框架(pretrained_init_configuration)本地、community 都没找到的情况下，是否可以直接默认去 huggingface 上面去找呢？
4. 是否需要针对于调整添加一些单测来保证稳定性呢？

sijunhe committed:
unit test pass

sijunhe committed:
fix styles

sijunhe committed:
Merge remote-tracking branch 'origin/develop' into hf_hub_integration

sijunhe committed:
Merge branch 'hf_hub_integration' of https://github.com/PaddlePaddle/PaddleNLP into hf_hub_integration

sijunhe commented:
> 你的这个 pr 我觉得非常棒，看起来很令人兴奋，不过我除了留的 comment，还有几点想跟你讨论：
> 
> 1. 你现在是直接在老版本的from_pretrained（新版本为`from_pretrained_v2`）上面做的，可是后面的主流模型都会迁移到新版本上，所以我感觉这部分的调整是否直接在新的接口上来完成是否比较好？
> 2. `from_hf_hub`的核心逻辑应该是数据源的问题，那是否可以把这部分的逻辑抽离成一个单独的函数来完成下载的这个工作呢？
> 3. 是否不用添加`from_hf_hub`这个参数，当在框架(pretrained_init_configuration)本地、community 都没找到的情况下，是否可以直接默认去 huggingface 上面去找呢？
> 4. 是否需要针对于调整添加一些单测来保证稳定性呢？

1. added the logic in `from_pretrained_v2` as well, which is tested through the `bert` model
2. discussed offline that we need `from_hf_hub`
3. same as above
4. added integration test

sijunhe committed:
add todo

wj-Mcat committed:
Merge branch 'develop' into hf_hub_integration

sijunhe committed:
integration tests

sijunhe committed:
Merge branch 'hf_hub_integration' of https://github.com/PaddlePaddle/PaddleNLP into hf_hub_integration

sijunhe committed:
Merge branch 'develop' into hf_hub_integration

sijunhe committed:
Merge branch 'develop' into hf_hub_integration
